baseapp/auth/saml/serviceprovider.go

```diff
@@ -0,0 +1,253 @@
+	}
+
+	provider := &saml.ServiceProvider{
+		IDPMetadata: e,
+		Certificate: cert,
+		Key:         rsaKey,
+	}
+	sp := &ServiceProvider{
+		sp:           provider,
+		onError:      settings.OnError,
+		onLogin:      settings.OnLogin,
+		acsPath:      settings.AssertionConsumerServicePath,
+		metadataPath: settings.SPMetadataPath,
+		idStore:      settings.IDStore,
+	}
+
+	if sp.onError == nil {
+		sp.onError = DefaultErrorCallback
+	}
+
+	if sp.onLogin == nil {
+		sp.onLogin = DefaultLoginCallback
+	}
+
+	if sp.idStore == nil {
+		sp.idStore = cookieIDStore{}
+	}
+
+	return sp, nil
+}
+
+func DefaultErrorCallback(r *http.Request, err error) {
+	fmt.Println(err.Error())
+}
+
+func DefaultLoginCallback(w http.ResponseWriter, r *http.Request, resp *saml.Assertion) {
+	w.WriteHeader(http.StatusOK)
+}
+
+func getEntityFromMetadata(settings Settings) (*saml.EntityDescriptor, error) {
+	resp, err := http.Get(settings.IDPMetadataURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to download IDP metadata")
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+	descriptor, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to download IDP metadata")
+	}
+
+	entity := &saml.EntityDescriptor{}
+
+	if err := xml.Unmarshal(descriptor, entity); err != nil {
+		return nil, errors.Wrap(err, "could not parse returned metadata")
+	}
+
+	return entity, nil
+}
+
+func (s *ServiceProvider) getSAMLSettingsForRequest(r *http.Request) *saml.ServiceProvider {
+
+	//make a copy in case different requests have different host headers
+	newSP := *s.sp
+
+	u := url.URL{
+		Host:   r.Host,
+		Scheme: "http",
+	}
+
+	if r.TLS != nil {
+		u.Scheme = "https"
+	}
+
+	u.Path = s.metadataPath
+	newSP.MetadataURL = u
+	u.Path = s.acsPath
+	newSP.AcsURL = u
+
+	return &newSP
+}
+
+// DoAuth takes an http.ResponseWriter that has not been written to yet, and conducts and SP initiated login
+// If the flow proceeds correctly the user should be redirected to the handler provided by ACSHandler().
+func (s *ServiceProvider) DoAuth(w http.ResponseWriter, r *http.Request) {
+	sp := s.getSAMLSettingsForRequest(r)
+
+	request, err := sp.MakeAuthenticationRequest(sp.GetSSOBindingLocation(saml.HTTPRedirectBinding))
+	if err != nil {
+		s.onError(r, errors.Wrap(err, "failed to create authentication request"))
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	if err := s.idStore.StoreID(w, request.ID); err != nil {
+		s.onError(r, errors.Wrap(err, "failed to store SAML request id"))
+		http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		return
+	}
+
+	target := request.Redirect("")
+
+	http.Redirect(w, r, target.String(), http.StatusFound)
+}
+
+// ACSHandler returns an http.Handler which is capable of validating and processing SAML Responses.
+func (s *ServiceProvider) ACSHandler() http.Handler {
```

Comment:
What does `ACS` stand for?

Reply:
AssertionConsumerService which is where the SAML IDP will send the user with the assertion.